library/cipher_wrap.c

```diff
@@ -89,12 +89,16 @@ enum mbedtls_cipher_base_index {
 #if defined(MBEDTLS_CCM_C) && defined(MBEDTLS_CAMELLIA_C)
     MBEDTLS_CIPHER_BASE_INDEX_CCM_CAMELLIA,
 #endif
-#if defined(MBEDTLS_CHACHA20_C)
-    MBEDTLS_CIPHER_BASE_INDEX_CHACHA20_BASE,
+#if defined(MBEDTLS_CIPHER_HAVE_GCM_AES_VIA_LEGACY_OR_USE_PSA)
+    MBEDTLS_CIPHER_BASE_INDEX_GCM_AES,
 #endif
-#if defined(MBEDTLS_CHACHAPOLY_C)
-    MBEDTLS_CIPHER_BASE_INDEX_CHACHAPOLY_BASE,
+#if defined(MBEDTLS_GCM_C) && defined(MBEDTLS_ARIA_C)
+    MBEDTLS_CIPHER_BASE_INDEX_GCM_ARIA,
+#endif
+#if defined(MBEDTLS_GCM_C) && defined(MBEDTLS_CAMELLIA_C)
+    MBEDTLS_CIPHER_BASE_INDEX_GCM_CAMELLIA,
 #endif
+#if defined(MBEDTLS_CIPHER_C)
```

Comment:
I think this could be placed higher - just after Camellia. We don't want cipher light to expose GCM or CCM as part of its API - we just want it to provide the necessary building blocks for `gcm.c` and `ccm.c`.

Reply:
Right, when implementing this I kept CCM and GCM there because I knew that AEADs were the functionalities we want to keep with `CIPHER_LIGHT`. However I didn't realized that this kind of support from TLS point of view is achieved by means of "separate" functions like `mbedtls_cipher_aead_encrypt()` which does not go through `cipher_wrap` directly (`cipher_wrap` is still required but it only indirectly through CCM/GCM modules).
Long story short: thanks for pointing this out. I will fix it ;)

Reply:
> AEADs were the functionalities we want to keep with `CIPHER_LIGHT`

No they're not :) The functionalities we want to keep with `CIPHER_LIGHT` are those that _are used to build_ GCM and CCM, which is very different. We do _not_ want AEADs as part of the `CIPHER_LIGHT` API, because we don't need them. TLS calls `psa_aead_xxx()` which directly calls `mbedtls_gcm_xxx()` or `mbedtls_ccm_xxx()`, and nothing in the library tries to do GCM or CCM by using `mbedtls_cipher_xxx()` so we want them out of `CIPHER_LIGHT`.

(Said otherwise, we want in `CIPHER_LIGHT` the functions _that are called by_ `mbedtls_gcm_xxx()` / `mbedtls_ccm_xxx()` but we don't want `CIPHER_LIGHT` to call `mbedtls_gcm_xxx()` or `mbedtls_ccm_xxx()`.)

The design document in #8057 says so:
> This excludes:
> - [...]
> - the AEAD/KW API (both one-shot and streaming);

Reply:
Right. Initially I set guards in cipher module based on what I thought it was required, but I didn't check the TLS code. Once I did yesterday, I immediately realized that you were right ;)